File: 0047-permutations-ii/0047-permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    let answ = [];
    let count = {}
    for(let i = 0; i < nums.length; i++){
        if(count[nums[i]] == undefined){
            count[nums[i]] = 1
        }else{
            count[nums[i]]++
        }
    }
    backtrack([]);

    function backtrack(curr) {
        if(curr.length == nums.length){
            answ.push([...curr])
            return
        }
        for(let key in count){
            if(count[key] > 0){
                count[key]--
                backtrack([...curr, Number(key)])
                count[key]++
            }
        }
    }
    
    return answ;
};

File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> answ;
        helper(nums, answ, 0);
        return answ;
    }
private:
    void helper(vector<int>& nums, vector<vector<int>>& answ, int index) {
        if (index == nums.size()) {
            answ.push_back(nums);
            return;
        }        
        unordered_set<int> seen;      
        for (int i = index; i < nums.size(); i++) {
            if (seen.count(nums[i])) {
                continue;
            }
            seen.insert(nums[i]);
            swap(nums[index], nums[i]);
            helper(nums, answ, index + 1);
            swap(nums[index], nums[i]);
        }
    }
};